app/scripts/controllers/transactions/index.js

```diff
@@ -1835,6 +1656,252 @@ export default class TransactionController extends EventEmitter {
   //
   //           PRIVATE METHODS
   //
+
+  async _waitForTransactionFinished(txId) {
+    return new Promise((resolve) => {
+      this.txStateManager.once(`${txId}:finished`, (txMeta) => {
+        resolve(txMeta);
+      });
+    });
+  }
+
+  async _createTransaction(
+    txMethodType,
+    txParams,
+    origin,
+    transactionType,
+    sendFlowHistory = [],
+    actionId,
+    options,
+  ) {
+    if (
+      transactionType !== undefined &&
+      !VALID_UNAPPROVED_TRANSACTION_TYPES.includes(transactionType)
+    ) {
+      throw new Error(
+        `TransactionController - invalid transactionType value: ${transactionType}`,
+      );
+    }
+
+    // If a transaction is found with the same actionId, do not create a new speed-up transaction.
+    if (actionId) {
+      let existingTxMeta =
+        this.txStateManager.getTransactionWithActionId(actionId);
+      if (existingTxMeta) {
+        this.emit('newUnapprovedTx', existingTxMeta);
```

Comment:
Are we emitting this event only for the unit tests?

If so, we might want to think on how to refactor them, even if it's in a future PR.

Reply:
Didn't really like this as I wrote it. I've removed the event and am instead using a spy on the state manager.